File: example/native-android/app/src/main/kotlin/id/frak/example/android/HarnessModel.kt
package id.frak.example.android

import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.pm.PackageInfoCompat
import id.frak.sdk.Frak
import id.frak.sdk.FrakSdkVersion
import id.frak.sdk.core.FrakError
import id.frak.sdk.core.FrakResult
import id.frak.sdk.core.ProductDetails
import id.frak.sdk.rewards.RewardRequest
import id.frak.sdk.sharing.SharingRequest
import id.frak.sdk.ui.FrakSharing
import id.frak.sdk.ui.SharingResult
import java.util.Date

/** Distinct from the SDK tags (Frak, FrakSharing) so a run can separate app from SDK. */
private const val HARNESS_TAG: String = "FrakHarness"

/** Bounded so a long test session cannot grow the log without limit. */
private const val MAX_LOG_ENTRIES = 300

/** Every SDK call the harness makes, plus the state the three tabs render. */
class HarnessModel(
    private val appContext: Context,
    private val sharing: FrakSharing,
    val activeEnvironment: HarnessEnvironment,
) {
    val logs = mutableStateListOf<LogEntry>()
    val debugRows = mutableStateListOf<DebugRow>()

    var catalogReward by mutableStateOf<CatalogRewardLookup>(CatalogRewardLookup.Loading)
        private set
    var sdkState by mutableStateOf<SdkState>(SdkState.Checking)
        private set
    var walletInstalled by mutableStateOf<Boolean?>(null)
        private set
    var currencyCode by mutableStateOf(FALLBACK_CURRENCY_CODE)
        private set
    var isDebugRefreshing by mutableStateOf(false)
        private set
    var selectedEnvironment by mutableStateOf(activeEnvironment)
        private set

    val buildLabel: String =
        runCatching {
            val info = appContext.packageManager.getPackageInfo(appContext.packageName, 0)
            "${info.versionName} (${PackageInfoCompat.getLongVersionCode(info)})"
        }.getOrDefault("unknown")

    val logExport: String get() = logs.asReversed().joinToString("\n") { it.exportLine }

    val debugExport: String get() = debugRows.joinToString("\n") { "${it.label}: ${it.value}" }

    suspend fun start() {
        addLog(
            "Frak.initialize called for merchant ${activeEnvironment.merchantId} " +
                "(${activeEnvironment.label})",
            LogType.INFO,
        )
        val installed = Frak.client.appLink.isFrakAppInstalled()
        walletInstalled = installed
        addLog("Frak wallet app installed: $installed", LogType.INFO)
        resolveConfig()
        loadCatalogReward()
        refreshDebugInfo(log = false)
    }

    /**
     * Persists only. Switching in place would leave the sharing sheet's pooled web view on the
     * previous stage's wallet origin, and re-initializing double-registers the deep-link observer.
     */
    fun selectEnvironment(environment: HarnessEnvironment) {
        selectedEnvironment = environment
        HarnessEnvironmentStore.write(appContext, environment)
        if (environment == activeEnvironment) {
            addLog("Environment stays ${environment.label} on the next launch.", LogType.INFO)
            return
        }
        addLog("Environment set to ${environment.label}. Restart the app to apply it.", LogType.INFO)
    }

    /** Share #1: no products at all — the link falls back to the merchant homepage. */
    fun shareStore() {
        addLog("Opening the share sheet for the whole store...", LogType.INFO)
        sharing.present(
            SharingRequest {
                targetInteraction = "purchase"
                placement = "home"
            },
        )
    }

    /** Share #2: exactly one product, scoped and illustrated. */
    fun shareProduct(product: ProductItem) {
        addLog("Opening the share sheet for '${product.title}'...", LogType.INFO)
        sharing.present(
            SharingRequest {
                products = listOf(sharingProduct(product))
                // Matches the rewards.best call below.
                targetInteraction = "purchase"
                placement = "product-page"
            },
        )
    }

    /** Share #3: the whole catalog, so the sheet renders several illustrated product cards. */
    fun shareCollection() {
        addLog("Opening the share sheet for ${sampleProducts.size} products...", LogType.INFO)
        sharing.present(
            SharingRequest {
                // Products carry their own links, so the shared URL has to be stated: without it
                // the first product's link would win and the recipient would miss the collection.
                link = "$STORE_LINK/collections/best-sellers"
                products = sampleProducts.map(::sharingProduct)
                targetInteraction = "purchase"
                placement = "category-page"
            },
        )
    }

    fun logSharingResult(result: SharingResult) {
        when (result) {
            // Android cannot see what the user picked: NativeShare returns startActivity().isSuccess,
            // so this fires when the chooser opens.
            is SharingResult.Shared -> addLog("Share chooser opened for: ${result.link}", LogType.INFO)

            is SharingResult.Copied -> addLog("Reward link copied to clipboard: ${result.link}", LogType.SUCCESS)

            SharingResult.InstallStarted -> addLog("Wallet install flow started by the sharing sheet.", LogType.INFO)

            SharingResult.WalletOpened -> addLog("Wallet opened directly; identity handed off.", LogType.SUCCESS)

            SharingResult.Dismissed -> addLog("Sharing sheet dismissed by user.", LogType.INFO)

            is SharingResult.Failed -> addLog("Sharing failed: ${result.error.message}", LogType.ERROR)
        }
    }

    /**
     * Reports arrival at the activity, which is all this callback can honestly observe. Whether the
     * SDK tracked it is a separate question that only Debug info answers, so this is not a SUCCESS.
     */
    fun logInboundIntent(url: String) {
        addLog("Inbound link reached the app: $url", LogType.INFO)
        addLog("Open Debug to confirm the SDK picked up the referral.", LogType.INFO)
    }

    /** The only place in the harness that calls `handleReferral` directly. */
    suspend fun simulateDeepLink() {
        val testUrl = "https://example-merchant.com/product?fCtx=test_referral_token_android_9988"
        addLog("Simulating an inbound referral link...", LogType.INFO)
        try {
            val handled = Frak.client.appLink.handleReferral(testUrl)
            addLog(
                if (handled) {
                    "Referral recognized — this visit is attributed to the sharer."
                } else {
                    "That link carried no Frak referral."
                },
                if (handled) LogType.SUCCESS else LogType.INFO,
            )
        } catch (error: FrakError) {
            addLog("appLink.handleReferral(...) failed: ${error.message}", LogType.ERROR)
        }
    }

    suspend fun completeOrder() {
        val orderId = "ord_${System.currentTimeMillis()}"
        addLog("Completing order $orderId (${formatCents(SAMPLE_ORDER_TOTAL_CENTS, currencyCode)})...", LogType.INFO)
        when (
            val result =
                Frak.client.tracking.purchase(
                    customerId = SAMPLE_CUSTOMER_ID,
                    orderId = orderId,
                    token = SAMPLE_CHECKOUT_TOKEN,
                )
        ) {
            is FrakResult.Success -> addLog("Order $orderId sent to Frak.", LogType.SUCCESS)
            is FrakResult.Failure -> addLog("Order $orderId tracking failed: ${result.error.message}", LogType.ERROR)
        }
    }

    fun clearLogs() {
        logs.clear()
        addLog("Log cleared.", LogType.INFO)
    }

    private suspend fun resolveConfig() {
        try {
            val resolved = Frak.client.config.resolve()
            currencyCode = resolved.currency?.wireValue ?: FALLBACK_CURRENCY_CODE
            sdkState = SdkState.Connected
            addLog("Merchant config resolved: ${resolved.name} (${resolved.domain})", LogType.SUCCESS)
        } catch (error: FrakError) {
            sdkState = SdkState.Failed(error.message.orEmpty())
            addLog("Config resolve failed: ${error.message}", LogType.ERROR)
        }
    }

    /** One `rewards.best` call for the whole visible catalog, not one per row. */
    private suspend fun loadCatalogReward() {
        catalogReward =
            try {
                val best =
                    Frak.client.rewards.best(
                        RewardRequest {
                            // Matches SharingRequest.targetInteraction used by shareProduct.
                            targetInteraction = "purchase"
                            products =
                                sampleProducts.map { product ->
                                    ProductDetails {
                                        productId = product.id
                                        name = product.title
                                    }
                                }
                        },
                    )
                if (best != null) {
                    addLog("Catalog reward: ${best.formatted}", LogType.SUCCESS)
                    CatalogRewardLookup.Loaded(best)
                } else {
                    addLog("No campaign matched the catalog.", LogType.INFO)
                    CatalogRewardLookup.NoActiveReward
                }
            } catch (error: FrakError) {
                addLog("Catalog reward lookup failed: ${error.message}", LogType.ERROR)
                CatalogRewardLookup.Failed
            }
    }

    /**
     * Every wiring fact the SDK can answer for, in one snapshot: the configured merchant id next
     * to the one the backend resolved, the identity events are attributed to, and the origins the
     * calls actually go to.
     */
    suspend fun refreshDebugInfo(log: Boolean) {
        isDebugRefreshing = true
        val rows = mutableListOf<DebugRow>()
        rows += DebugRow("App build", buildLabel)
        rows += DebugRow("SDK version", FrakSdkVersion.CURRENT)
        rows += DebugRow("Harness environment", activeEnvironment.label)
        rows += DebugRow("Configured merchant id", activeEnvironment.merchantId)

        val client = Frak.clientOrNull
        if (client == null) {
            rows += DebugRow("Client", "not initialized")
            replaceDebugRows(rows)
            isDebugRefreshing = false
            if (log) addLog("SDK debug info: client not initialized.", LogType.ERROR)
            return
        }

        val environment = client.environment
        rows += DebugRow("Environment", environment::class.simpleName ?: "custom")
        rows += DebugRow("Wallet origin", environment.wallet)
        rows += DebugRow("Backend origin", environment.backend)
        rows += DebugRow("Wallet package id", environment.walletPackageId)

        val installed = client.appLink.isFrakAppInstalled()
        walletInstalled = installed
        rows += DebugRow("Wallet app installed", installed.toString())
        rows += DebugRow("Tracking enabled", client.isTrackingEnabled().toString())
        rows += DebugRow("Anonymous id", client.anonymousId() ?: "none (tracking off or key refused)")

        try {
            val resolved = client.config.resolve()
            currencyCode = resolved.currency?.wireValue ?: FALLBACK_CURRENCY_CODE
            sdkState = SdkState.Connected
            rows += DebugRow("Resolved merchant id", resolved.merchantId)
            rows += DebugRow("Merchant name", resolved.displayName)
            rows += DebugRow("Merchant domain", resolved.domain)
            rows += DebugRow("Currency", resolved.currency?.wireValue ?: "unset")
            rows += DebugRow("Language", resolved.lang?.wireValue ?: "unset")
            val placements = resolved.sdkConfig?.placements.orEmpty()
            rows += DebugRow("Configured placements", placements.keys.joinToString().ifEmpty { "none" })
        } catch (error: FrakError) {
            sdkState = SdkState.Failed(error.message.orEmpty())
            rows += DebugRow("Resolved config", "failed: ${error.message}")
        }

        replaceDebugRows(rows)
        isDebugRefreshing = false
        if (log) addLog("SDK debug info refreshed (${rows.size} fields).", LogType.SUCCESS)
    }

    private fun replaceDebugRows(rows: List<DebugRow>) {
        debugRows.clear()
        debugRows.addAll(rows)
    }

    fun addLog(
        message: String,
        type: LogType,
    ) {
        logs.add(0, LogEntry(LOG_TIME_FORMAT.format(Date()), message, type))
        if (logs.size > MAX_LOG_ENTRIES) {
            logs.removeRange(MAX_LOG_ENTRIES, logs.size)
        }
        // Mirrored so a device run is greppable from `adb logcat` instead of read off the screen.
        when (type) {
            LogType.ERROR -> Log.e(HARNESS_TAG, message)
            else -> Log.i(HARNESS_TAG, message)
        }
    }
}

File: example/native-ios/Sources/FrakExampleiOSApp/HarnessModel.swift
import Foundation
import FrakSDK
import FrakSDKUI
import SwiftUI
import os

/// Every SDK call the harness makes, plus the state the three tabs render.
@MainActor
final class HarnessModel: ObservableObject {
    /// Bounded so a long test session cannot grow the log without limit.
    private static let maxLogEntries = 300

    /// Distinct from the SDK's own subsystem so a run can separate app from SDK.
    private let logger = Logger(subsystem: "id.frak.example.ios", category: "FrakHarness")

    @Published private(set) var logs: [LogEntry] = []
    @Published private(set) var debugRows: [DebugRow] = []
    @Published private(set) var catalogReward: CatalogRewardLookup = .loading
    @Published private(set) var sdkState: SdkState = .checking
    @Published private(set) var walletInstalled: Bool?
    @Published private(set) var currencyCode = fallbackCurrencyCode
    @Published var isDebugRefreshing = false
    @Published var selectedEnvironment = HarnessEnvironmentStore.read()
    @Published var isSharingPresented = false
    @Published var pendingSharingRequest = SharingRequest()

    var pendingRestart: Bool { selectedEnvironment != activeEnvironment }

    var logExport: String {
        logs.reversed().map(\.exportLine).joined(separator: "\n")
    }

    var debugExport: String {
        debugRows.map { "\($0.label): \($0.value)" }.joined(separator: "\n")
    }

    func start() async {
        addLog(
            "Frak.initialize called for merchant \(activeEnvironment.merchantId) "
                + "(\(activeEnvironment.label))",
            type: .info
        )
        await checkWalletInstalled()
        await resolveConfig()
        await loadCatalogReward()
        await refreshDebugInfo(log: false)
    }

    /// Persists only. Switching in place would leave the sharing sheet's pooled web view on the
    /// previous stage's wallet origin, and `Frak.shutdown()` does not stop every background task.
    func selectEnvironment(_ environment: HarnessEnvironment) {
        selectedEnvironment = environment
        HarnessEnvironmentStore.write(environment)
        guard environment != activeEnvironment else {
            addLog("Environment stays \(environment.label) on the next launch.", type: .info)
            return
        }
        addLog("Environment set to \(environment.label). Relaunch the app to apply it.", type: .info)
    }

    /// Share #1: no products at all — the link falls back to the merchant homepage.
    func shareStore() {
        addLog("Opening the share sheet for the whole store...", type: .info)
        pendingSharingRequest = SharingRequest(
            targetInteraction: "purchase",
            placement: "home"
        )
        isSharingPresented = true
    }

    /// Share #2: exactly one product, scoped and illustrated.
    func shareProduct(_ product: ProductItem) {
        addLog("Opening the share sheet for '\(product.title)'...", type: .info)
        pendingSharingRequest = SharingRequest(
            products: [sharingProduct(product)],
            // Matches the rewards.best call below.
            targetInteraction: "purchase",
            placement: "product-page"
        )
        isSharingPresented = true
    }

    /// Share #3: the whole catalog, so the sheet renders several illustrated product cards.
    func shareCollection() {
        addLog("Opening the share sheet for \(sampleProducts.count) products...", type: .info)
        pendingSharingRequest = SharingRequest(
            // Products carry their own links, so the shared URL has to be stated: without it the
            // first product's link would win and the recipient would miss the collection.
            link: "\(storeLink)/collections/best-sellers",
            products: sampleProducts.map(sharingProduct),
            targetInteraction: "purchase",
            placement: "category-page"
        )
        isSharingPresented = true
    }

    func handleSharingResult(_ result: SharingResult) {
        switch result {
        case .shared(let link):
            addLog("Reward link shared: \(link)", type: .success)
        case .copied(let link):
            addLog("Reward link copied to clipboard: \(link)", type: .success)
        case .installStarted:
            addLog("Wallet install flow started by the sharing sheet.", type: .info)
        case .walletOpened:
            addLog("Wallet opened with the sharing sheet's identity handoff.", type: .success)
        case .dismissed:
            addLog("Sharing sheet dismissed by user.", type: .info)
        case .failed(let error):
            addLog("Sharing failed: \(error.localizedDescription)", type: .error)
        }
    }

    /// Both the deep-link simulator button and `.onOpenURL` delivery funnel through here.
    func handleInboundURL(_ url: URL) async {
        addLog("Inbound link reached the app: \(url)", type: .info)
        do {
            let client = try Frak.client
            let hadReferral = await client.appLink.handleReferral(url)
            addLog(
                hadReferral
                    ? "Referral recognized — this visit is attributed to the sharer."
                    : "That link carried no Frak referral.",
                type: hadReferral ? .success : .info
            )
        } catch {
            addLog("Frak.client unavailable: \(error.localizedDescription)", type: .error)
        }
    }

    func simulateDeepLink() async {
        let raw = "https://example-merchant.com/product?fCtx=test_referral_token_ios_9988"
        guard let url = URL(string: raw) else { return }
        addLog("Simulating an inbound referral link...", type: .info)
        await handleInboundURL(url)
    }

    func completeOrder() async {
        let orderId = "ord_\(Int(Date().timeIntervalSince1970))"
        addLog(
            "Completing order \(orderId) (\(formatCents(sampleOrderTotalCents, currencyCode: currencyCode)))...",
            type: .info
        )
        // tracking.purchase has no amount parameter; the total above is display-only.
        let result = await client()?.tracking.purchase(
            customerId: sampleCustomerId,
            orderId: orderId,
            token: sampleCheckoutToken
        )
        switch result {
        case .success:
            addLog("Order \(orderId) sent to Frak.", type: .success)
        case .failure(let error):
            addLog("Order \(orderId) tracking failed: \(error.localizedDescription)", type: .error)
        case nil:
            addLog("Frak.client unavailable — order \(orderId) not tracked.", type: .error)
        }
    }

    func clearLogs() {
        logs.removeAll()
        addLog("Log cleared.", type: .info)
    }

    private func checkWalletInstalled() async {
        guard let client = client() else {
            addLog("Frak.client unavailable — skipping wallet-installed check.", type: .error)
            return
        }
        let installed = await client.appLink.isFrakAppInstalled()
        walletInstalled = installed
        addLog("Frak wallet app installed: \(installed)", type: .info)
    }

    private func resolveConfig() async {
        guard let client = client() else {
            sdkState = .failed("client unavailable")
            addLog("Frak.client unavailable — skipping config resolve.", type: .error)
            return
        }
        do {
            let resolved = try await client.config.resolve()
            currencyCode = resolved.currency?.rawValue ?? fallbackCurrencyCode
            sdkState = .connected
            addLog("Merchant config resolved: \(resolved.name) (\(resolved.domain))", type: .success)
        } catch {
            sdkState = .failed(error.localizedDescription)
            addLog("Config resolve failed: \(error.localizedDescription)", type: .error)
        }
    }

    /// One `rewards.best(...)` call for the whole visible catalog, not one per product.
    private func loadCatalogReward() async {
        guard let client = client() else {
            addLog("Frak.client unavailable — skipping reward lookup.", type: .error)
            catalogReward = .failed
            return
        }
        do {
            let best = try await client.rewards.best(
                RewardRequest(
                    // Matches SharingRequest.targetInteraction used by shareProduct.
                    targetInteraction: "purchase",
                    products: sampleProducts.map { ProductDetails(productId: $0.id, name: $0.title) }
                )
            )
            if let best {
                catalogReward = .loaded(best)
                addLog("Catalog reward: \(best.formatted)", type: .success)
            } else {
                catalogReward = .noActiveReward
                addLog("No campaign matched the catalog.", type: .info)
            }
        } catch {
            addLog("Catalog reward lookup failed: \(error.localizedDescription)", type: .error)
            catalogReward = .failed
        }
    }

    /// Every wiring fact the SDK can answer for, in one snapshot: the configured merchant id next
    /// to the one the backend resolved, the identity events are attributed to, and the origins the
    /// calls actually go to.
    func refreshDebugInfo(log: Bool) async {
        isDebugRefreshing = true
        defer { isDebugRefreshing = false }

        var rows = [
            DebugRow(label: "App build", value: harnessBuildLabel),
            DebugRow(label: "SDK version", value: FrakSDKVersion.current),
            DebugRow(label: "Harness environment", value: activeEnvironment.label),
            DebugRow(label: "Configured merchant id", value: activeEnvironment.merchantId),
            DebugRow(label: "Bundle id", value: Bundle.main.bundleIdentifier ?? "unknown"),
        ]

        guard let client = client() else {
            rows.append(DebugRow(label: "Client", value: "not initialized"))
            debugRows = rows
            if log { addLog("SDK debug info: client not initialized.", type: .error) }
            return
        }

        let environment = client.environment
        rows.append(DebugRow(label: "Environment", value: "\(environment)"))
        rows.append(DebugRow(label: "Wallet origin", value: environment.wallet))
        rows.append(DebugRow(label: "Backend origin", value: environment.backend))
        rows.append(DebugRow(label: "Wallet scheme", value: environment.walletScheme))

        let installed = await client.appLink.isFrakAppInstalled()
        walletInstalled = installed
        rows.append(DebugRow(label: "Wallet app installed", value: "\(installed)"))
        rows.append(DebugRow(label: "Tracking enabled", value: "\(await client.isTrackingEnabled())"))
        rows.append(
            DebugRow(
                label: "Anonymous id",
                value: await client.anonymousId ?? "none (tracking off or key refused)"
            )
        )

        do {
            let resolved = try await client.config.resolve()
            currencyCode = resolved.currency?.rawValue ?? fallbackCurrencyCode
            sdkState = .connected
            rows.append(DebugRow(label: "Resolved merchant id", value: resolved.merchantId))
            rows.append(DebugRow(label: "Merchant name", value: resolved.displayName))
            rows.append(DebugRow(label: "Merchant domain", value: resolved.domain))
            rows.append(DebugRow(label: "Currency", value: resolved.currency?.rawValue ?? "unset"))
            rows.append(DebugRow(label: "Language", value: resolved.lang?.rawValue ?? "unset"))
            let placements = resolved.sdkConfig?.placements.keys.sorted() ?? []
            rows.append(
                DebugRow(
                    label: "Configured placements",
                    value: placements.isEmpty ? "none" : placements.joined(separator: ", ")
                )
            )
        } catch {
            sdkState = .failed(error.localizedDescription)
            rows.append(DebugRow(label: "Resolved config", value: "failed: \(error.localizedDescription)"))
        }

        debugRows = rows
        if log { addLog("SDK debug info refreshed (\(rows.count) fields).", type: .success) }
    }

    /// Adapts `Frak.client`'s throw to a plain optional, in one place.
    private func client() -> FrakClient? {
        try? Frak.client
    }

    func addLog(_ message: String, type: LogEntry.LogType) {
        let timestamp = logTimeFormatter.string(from: Date())
        logs.insert(LogEntry(timestamp: timestamp, message: message, type: type), at: 0)
        if logs.count > Self.maxLogEntries {
            logs.removeLast(logs.count - Self.maxLogEntries)
        }
        // Mirrored so a device run is greppable from the console instead of read off the screen.
        switch type {
        case .error: logger.error("\(message, privacy: .public)")
        default: logger.info("\(message, privacy: .public)")
        }
    }
}
